pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
+// syncNetworkDynamicMacAddress saves the dynamic mac address observed from the the VMI status to vm
+func (h *VMNetworkController) syncNetworkDynamicMacAddress(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
 	if err != nil {
 		return vmi, err
 	}
 
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	vmCopy := vm.DeepCopy()
+	vmiInterfaces := getVmiInterfaceMacs(vmi)
+	updated := false
+	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmIface.Name == "" || vmIface.MacAddress != "" {
+			continue
+		}
+		if macAddress := vmiInterfaces[vmIface.Name]; macAddress != "" {
+			// only backfill the network mac address when it is empty
+			logrus.Infof("set VM %s network %s macAddress to %s", vm.Name, vmIface.Name, macAddress)
+			vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces[i].MacAddress = macAddress
+			updated = true
+		}
+	}
+	if !updated {
+		return vmi, nil
+	}
+	*retry = true
+	if _, err := h.vmClient.Update(vmCopy); err != nil {
+		return vmi, fmt.Errorf("failed to update vmi mac to vm: %w", err)
+	}
+
 	return vmi, nil
 }
 
-func (h *VMNetworkController) updateVMDefaultNetworkMacAddress(vmi *kubevirtv1.VirtualMachineInstance) error {
-	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
-	if err != nil {
-		return err
+func getVMInterfaceMacs(vm *kubevirtv1.VirtualMachine) map[string]string {
+	vmInterfaceMacs := make(map[string]string, len(vm.Spec.Template.Spec.Domain.Devices.Interfaces))
+	for _, vmIface := range vm.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmIface.MacAddress != "" {
+			vmInterfaceMacs[vmIface.Name] = vmIface.MacAddress
+		}
 	}
+	return vmInterfaceMacs
+}
 
-	vmCopy := vm.DeepCopy()
-	var length = len(vmi.Status.Interfaces)
-	var vmiInterfaces = make(map[string]string, length)
+func getVmiInterfaceMacs(vmi *kubevirtv1.VirtualMachineInstance) map[string]string {
+	var vmiInterfaces = make(map[string]string, len(vmi.Status.Interfaces))
 	for _, iface := range vmi.Status.Interfaces {
-		if iface.MAC != "" && iface.Name != "" {
+		if iface.Name != "" && iface.MAC != "" {
 			vmiInterfaces[iface.Name] = iface.MAC
 		}
 	}
+	return vmiInterfaces
+}
 
-	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
-		macAddress, ok := vmiInterfaces[vmIface.Name]
-		// only set the network mac address when it has no existing value
-		if ok && vmIface.MacAddress == "" {
-			logrus.Debugf("set VM %s management network %s macAddress to %s", vm.Name, vmIface.Name, macAddress)
-			vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces[i].MacAddress = macAddress
-		}
+func (h *VMNetworkController) syncControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
```

Comment:
Seems like we keep repeating this call in all the private methods. Is the intention here to ensure we always work with the latest vm? Otherwise, we can just call it once inside `SyncMacAddressAndControllerRevision()`, right?

Reply:
besides harvester contrller backfills the MAC;  user may edit VM at the meantime (e.g. set A MAC), we can't use VM spce to overwirte controllerrevision wholely or vice versa;  strictly, original code does not handle this race case well. https://github.com/harvester/harvester/pull/7402#issue-2798817503  test case 5

harveser only backfill MAC from VMI to VM&CR when the interfaces are there and MACs are nil